Add resource semantic conventions for operating systems (#693)
open-telemetry/opentelemetry-specification: specification/resource/semantic_conventions/README.md

## Environment

Attributes defining a running environment (e.g. Operating System, Cloud, Data Center).

### Operating System

**type:** `os`

**Description**: The operating system (OS) on which the process represented by this resource is running.

In case of virtualized environments, this is the operating system as it is observed by the process, i.e., the virtualized guest rather than the underlying host.

| Attribute  | Type | Description  | Example  | Required |
|---|---|---|---|---|
| `os.type` | string | The operating system type. | `"WINDOWS"` | Yes |
| `os.description` | string | Human readable (not intended to be parsed) OS version information, like e.g. reported by `ver` or `lsb_release -a` commands. | `"Microsoft Windows [Version 10.0.18363.778]"`<br>`"Ubuntu 18.04.1 LTS"` | No |

> `os.type` MUST be one of the following or, if none of the listed values apply, a custom value:

Reviewer: This can be downgraded to a recommendation. There is no incompatibility worry, and if we are saying a custom type is also acceptable there is no real normative requirement statement here.

```suggestion
`os.type` SHOULD be one of the following. If none of the listed values apply, a custom value best describing the family the operating system belongs to CAN be used.
```

Author: There's a fine difference as with the current wording you'd have to use one of the listed values if one is applicable to the OS you're reporting, whereas with the suggested wording you could use any arbitrary value even if there would be a well-defined one for that OS (e.g. `MS-WIN` instead of `WINDOWS`). This would make it unreliable for backends that want to make use of at least the well-known values for anything other than just showing the raw string in the UI.
I like the last part though and will add it right away.

Reviewer: > with the suggested wording you could use any arbitrary value even if there would be a well-defined one for that OS (e.g. MS-WIN instead of WINDOWS).

It is true that you _could_, it is not true that you _should_. If the reader is following the defined meaning of SHOULD as set out in [RFC 2119](https://tools.ietf.org/html/rfc2119) (as laid out in [our notation conventions](https://github.com/open-telemetry/opentelemetry-specification#notation-conventions-and-compliance)) they need to justify not using a predefined value:

> **SHOULD**   This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

If we are saying there exist valid reasons in particular circumstances that one of the value is not used (i.e. "none of the listed values apply"), it would seem to follow this level of normative directive is appropriate.

Conversely, with regard to using the rigid requirement of MUST, RFC 2119 requires these imperative to be used sparing:

> [Imperatives of the type defined in this memo must be used with care and sparingly.  In particular, they MUST only be used where it is actually required for interoperation or to limit behavior which has potential for causing harm (e.g., limiting retransmisssions)  For example, they must not be used to try to impose a particular method on implementors where the method is not required for interoperability.](https://tools.ietf.org/html/rfc2119#section-6)

If the specification states that a custom value is also valid, it implies that any handler of this data needs to also support custom values. Like it was pointed out, this support might not be ideal and might include just displaying the raw value, but it will need to exist. Give the backend is going to have to support values outside of this list, it follows there is no hard requirement for interoperation. Alternate values will not break things.

If alternate values are indeed a concern of interoperation (or maybe a security concern(?)) we should provide more detail about the form of the allowed custom values to address this.

Author: Since custom values are allowed as well, using values other than the pre-defined ones will of course not cause harm or pose a security concern. I'm fine with making it a SHOULD then since it's obvious that using any custom value can't result in the same "automatic" handling/detection if it's not defined among the well-known ones.

| Value  | Description |
|---|---|
| `WINDOWS` | Microsoft Windows |

> | `LINUX` | Linux |

Reviewer: What about Unix itself?

Author: That would be covered by the Unix systems below although we could add more there (e.g. the whole BSD line-up) or a generic `UNIX` (or to be discourage from using it when there are closer matching alternatives) `OTHER_UNIX`. WDYT?

Reviewer: Yeah, I was considering adding Unix as a catch all. However, you make a good point that it would overlap with what we have already.

If we want to be more specific than Unix, we should probably add the `BSD` family: `FreeBSD`, `NetBSD`, and `OpenBSD`?

Author: added in 2fe2218

Reviewer 2: I'm not very knowledgeable about BSD, but I think adding the three BSDs separately would make sense. Wikipedia lists only four of them. https://en.wikipedia.org/wiki/Comparison_of_BSD_operating_systems

Author: Me neither. Someone more knowledgeable please help out :-)

## Environment

Attributes defining a running environment (e.g. Operating System, Cloud, Data Center).

### Operating System

**type:** `os`

**Description**: The operating system (OS) on which the process represented by this resource is running.

In case of virtualized environments, this is the operating system as it is observed by the process, i.e., the virtualized guest rather than the underlying host.

| Attribute  | Type | Description  | Example  | Required |
|---|---|---|---|---|
| `os.type` | string | The operating system type. | `"WINDOWS"` | Yes |
| `os.description` | string | Human readable (not intended to be parsed) OS version information, like e.g. reported by `ver` or `lsb_release -a` commands. | `"Microsoft Windows [Version 10.0.18363.778]"`<br>`"Ubuntu 18.04.1 LTS"` | No |

`os.type` SHOULD be set to one of the values listed below.
If none of the listed values apply, a custom value best describing the family the operating system belongs to CAN be used.

| Value  | Description |
|---|---|
| `WINDOWS` | Microsoft Windows |
| `LINUX` | Linux |
| `DARWIN` | Apple Darwin |

> | `BSD` | BSD (Berkeley Software Distribution) family |

Author: Alternatively:

```suggestion
| `FREEBSD` | FreeBSD |
| `NETBSD` | NetBSD|
| `OPENBSD` | OpenBSD |
| `DRAGONFLY_BSD` | DragonFly BSD |
```

Reviewer: This seems reasonable to me :+1:

| `WINDOWS` | Microsoft Windows |
| `LINUX` | Linux |
| `DARWIN` | Apple Darwin |
| `FREEBSD` | FreeBSD |
| `NETBSD` | NetBSD|
| `OPENBSD` | OpenBSD |

> | `DRAGONFLY_BSD` | DragonFly BSD |

Reviewer 2: Why is there an underscore only here? Only because the "Marketing name" contains a space? I'd say, either use an underscore everywhere or nowhere.

```suggestion
| `DRAGONFLYBSD` | DragonFly BSD |
```
